### PR TITLE
Feature improvements for running at scale

### PR DESCRIPTION
**Pull request by bvanessen**

### CODE REVIEW

File: src/callbacks/ltfb.cpp [hunk 1]
```diff
@@ -314,18 +314,21 @@ EvalType evaluate(model& m, const std::string& metric_name) {
 
   // Mark the data store as loading - Note that this is a temporary fix
   // for the current use of the tournament
-  m.mark_data_store_explicitly_loading(execution_mode::validation);
+  // m.mark_data_store_explicitly_loading(execution_mode::validation); /// @todo RESTORE
+  m.mark_data_store_explicitly_loading(execution_mode::testing); /// @todo REMOVE
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:318]:
```suggestion
  m.mark_data_store_explicitly_loading(execution_mode::validation);
```

File: src/callbacks/ltfb.cpp [hunk 2]
```diff
@@ -314,18 +314,21 @@ EvalType evaluate(model& m, const std::string& metric_name) {
...
 
-  // Evaluate model on validation set
-  c.get_trainer().evaluate(&m, execution_mode::validation);
+  // Evaluate model on testing set
+  // c.get_trainer().evaluate(&m, execution_mode::validation); /// @todo RESTORE
+  c.get_trainer().evaluate(&m, execution_mode::testing); /// @todo REMOVE
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:322]:
```suggestion
  c.get_trainer().evaluate(&m, execution_mode::validation);
```

File: src/callbacks/ltfb.cpp [hunk 3]
```diff
@@ -314,18 +314,21 @@ EvalType evaluate(model& m, const std::string& metric_name) {
...
 
   // Get metric value
   bool found_metric = false;
   EvalType metric_value = 0;
   for (const auto& met : m.get_metrics()) {
     if (met->name() == metric_name) {
       found_metric = true;
-      metric_value = met->get_mean_value(execution_mode::validation);
+      // metric_value = met->get_mean_value(execution_mode::testing); /// @todo RESTORE
+      metric_value = met->get_mean_value(execution_mode::testing); /// @todo REMOVE
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:331]:
```suggestion
      metric_value = met->get_mean_value(execution_mode::validation);
```

File: src/callbacks/ltfb.cpp [hunk 4]
```diff
@@ -336,7 +339,8 @@ EvalType evaluate(model& m, const std::string& metric_name) {
 
   // Mark the data store as loaded - Note that this is a temporary fix
   // for the current use of the tournament
-  m.make_data_store_preloaded(execution_mode::validation);
+  // m.make_data_store_preloaded(execution_mode::validation); /// @todo RESTORE
+  m.make_data_store_preloaded(execution_mode::testing); /// @todo REMOVE
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:343]:
```suggestion
  m.make_data_store_preloaded(execution_mode::validation);
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
When we actually want to merge this into develop, make sure to revert LTFB to use the validation set.

File: src/callbacks/ltfb.cpp [hunk 5]
```diff
@@ -517,7 +517,9 @@ void ltfb::on_batch_begin(model *m) {
   El::Int tournament_winner = partner_trainer;
   if ((m_low_score_wins && local_score <= partner_score)
       || (!m_low_score_wins && local_score >= partner_score)
-      || (!std::isnan(local_score) && std::isnan(partner_score))) {
+      || (!std::isnan(local_score) && std::isnan(partner_score))
+      || (local_score != 0 && partner_score == 0) /** @todo REMOVE! */
+    ) {
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:522]:
```suggestion
      || (!std::isnan(local_score) && std::isnan(partner_score))) {
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Revert this hack before merging back into develop.

File: src/callbacks/ltfb.cpp [hunk 6]
```diff
@@ -461,7 +461,8 @@ void ltfb::on_batch_begin(model *m) {
   if (comm.am_world_master()) {
     std::cout << message_prefix + "evaluating local model...\n";
   }
-  const auto local_score = evaluate(*m, m_metric_name);
+  auto local_score = evaluate(*m, m_metric_name);
+  if (local_score == 0) { local_score = std::nan(""); } /// @todo REMOVE!
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:465]:
```suggestion
  const auto local_score = evaluate(*m, m_metric_name);
```

File: src/callbacks/ltfb.cpp [hunk 7]
```diff
@@ -510,14 +511,16 @@ void ltfb::on_batch_begin(model *m) {
   if (comm.am_world_master()) {
     std::cout << message_prefix + "evaluating partner model...\n";
   }
-  const auto& partner_score = evaluate(*m, m_metric_name);
+  auto partner_score = evaluate(*m, m_metric_name);
+  if (partner_score == 0) { partner_score = std::nan(""); } /// @todo REMOVE!
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:515]:
```suggestion
  const auto partner_score = evaluate(*m, m_metric_name);
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

**Comment by bvanessen** [comment]
@timmoon10 This PR has been rebased onto develop to get all of the new features.

File: src/data_readers/data_reader_smiles.cpp [hunk 8]
```diff
@@ -124,7 +138,7 @@ void smiles_data_reader::load() {
   resize_shuffled_indices();
 
   // Optionally run "poor man's" LTFB
-  if (opts->get_bool("ltfb")) {
+  if (opts->get_bool("ltfb") && get_role() != "test") {
```
Comment (timmoon10) [inline src/data_readers/data_reader_smiles.cpp:141]: Is this correct?

File: src/training_algorithms/sgd_training_algorithm.cpp [hunk 9]
```diff
@@ -88,12 +88,14 @@ void sgd_training_algorithm::train(sgd_execution_context& c,
     do_epoch_end_cbs(model);
 
     // Evaluate on validation set
-    auto key = c.get_trainer().check_and_build_execution_context(c, model, execution_mode::validation);
+    //auto key = c.get_trainer().check_and_build_execution_context(c, model, execution_mode::validation);
+    // Hack to use testing as validation in LTFB training, do not merge
+    auto key = c.get_trainer().check_and_build_execution_context(c, model, execution_mode::testing);
```
Comment (timmoon10) [inline src/training_algorithms/sgd_training_algorithm.cpp:93]:
```suggestion
    auto key = c.get_trainer().check_and_build_execution_context(c, model, execution_mode::validation);
```

File: src/training_algorithms/sgd_training_algorithm.cpp [hunk 10]
```diff
@@ -88,12 +88,14 @@ void sgd_training_algorithm::train(sgd_execution_context& c,
...
     auto& evaluation_context = static_cast<sgd_execution_context&>(c.get_trainer().get_execution_context(key));
     // Check to make sure that the model has a valid execution mode
     // before trying to do inference
-    if (model.is_execution_mode_valid(execution_mode::validation)) {
-      evaluate(evaluation_context, model, dc, execution_mode::validation);
+    if (model.is_execution_mode_valid(execution_mode::testing)) {
+      evaluate(evaluation_context, model, dc, execution_mode::testing);
```
Comment (timmoon10) [inline src/training_algorithms/sgd_training_algorithm.cpp:98]:
```suggestion
    if (model.is_execution_mode_valid(execution_mode::validation)) {
      evaluate(evaluation_context, model, dc, execution_mode::validation);
```

File: src/callbacks/ltfb.cpp [hunk 11]
```diff
@@ -456,7 +456,9 @@ void ltfb::on_batch_begin(model *m) {
   if (comm.am_world_master()) {
     std::cout << message_prefix + "evaluating local model...\n";
   }
-  const auto local_score = evaluate(*m, m_metric_name);
+  auto local_score = evaluate(*m, m_metric_name);
+  if (local_score == 0) { local_score = std::nan(""); } /// @todo REMOVE!
+  if (std::isinf(local_score)) { local_score = std::nan(""); } /// @todo REMOVE!
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:461]: It makes sense to me that inf scores are invalid, but zeros could be perfectly sensible, e.g. if we evaluate with categorical accuracy.
```suggestion
  auto local_score = evaluate(*m, m_metric_name);
  if (std::isinf(local_score)) { local_score = std::nan(""); }
```

File: src/callbacks/ltfb.cpp [hunk 12]
```diff
@@ -505,14 +507,17 @@ void ltfb::on_batch_begin(model *m) {
   if (comm.am_world_master()) {
     std::cout << message_prefix + "evaluating partner model...\n";
   }
-  const auto& partner_score = evaluate(*m, m_metric_name);
+  auto partner_score = evaluate(*m, m_metric_name);
+  if (partner_score == 0) { partner_score = std::nan(""); } /// @todo REMOVE!
+  if (std::isinf(partner_score)) { partner_score = std::nan(""); } /// @todo REMOVE!
```
Comment (timmoon10) [inline src/callbacks/ltfb.cpp:512]:
```suggestion
  auto partner_score = evaluate(*m, m_metric_name);
  if (std::isinf(partner_score)) { partner_score = std::nan(""); }
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: src/data_readers/data_reader_smiles.cpp (hunk 8 above)
Comment (bvanessen) [reply src/data_readers/data_reader_smiles.cpp:141]: Yes this is for the way it does LTFB.

**Comment by bvanessen** [comment]
@timmoon10 I merged your suggestions. Any other ones?

**Review by timmoon10 — APPROVED** [review]
LGTM

**Pull request merged**